v2/pkg/genruntime/configmaps/collector.go

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package configmaps
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"golang.org/x/exp/maps"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+// TODO: This is similar to secrets.Collector. If this is updated, that should likely be as well
+
+// Collector collects secret values and their associated genruntime.ConfigMapDestination's
+// and produces a merged set of v1.Secret's that can be written.
+type Collector struct {
+	configs   map[string]*v1.ConfigMap
+	namespace string
+	errors    []error
+}
+
+// NewCollector creates a new Collector for collecting multiple config map writes together
+func NewCollector(namespace string) *Collector {
+	return &Collector{
+		configs:   make(map[string]*v1.ConfigMap),
+		namespace: namespace,
+	}
+}
+
+// AddValue adds the dest and ConfigMapDestination pair to the collector. If another value has already
+// been added going to the same config map (but with a different key) the new key is merged into the
+// existing map.
+func (c *Collector) AddValue(dest *genruntime.ConfigMapDestination, value string) {
+	if dest == nil || value == "" {
+		return
+	}
+
+	existing, ok := c.configs[dest.Name]
+	if !ok {
+		existing = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      dest.Name,
+				Namespace: c.namespace,
+			},
+			Data: make(map[string]string),
+		}
+		c.configs[dest.Name] = existing
+	}
+
+	if _, ok = existing.Data[dest.Key]; ok {
+		c.errors = append(c.errors, errors.Errorf("key collision, two entries for key %s", dest.Key))
+		return
+	}
+	existing.Data[dest.Key] = value
+}
+
+// Values returns the set of Values that have been collected.
+func (c *Collector) Values() ([]*v1.ConfigMap, error) {
+	err := kerrors.NewAggregate(c.errors)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I'm still on the fence with this idiom - would prefer if the check was more like this

```suggestion
	if len(c.errors) > 0 {
		return nil, kerrors.NewAggregate(c.errors)
	}
```

Reply:
Ah I feel sorta the opposite. This idiom matches the `err := func(); err != nil` idiom that Go has with almost all functions. The `len` checking one on the other hand doesn't immediately look like every other error handling block/return. I took a look through https://github.com/kubernetes-sigs and see all 3 usages of the package:

`return errors.NewAggregate(errs)` was probably the most common.
There was a good number of `if len(errs) > 0 { return errors.NewAggregate(errs)`, and also a good number of `err := errors.NewAggregate(errs); err != nil` (this idiom).

If folks feel strongly we should change this, I'm fine doing so, but I do think that the `err := <x>` idiom is the standard Go err handling idiom and this matches that and so scans easier. We also use this pattern elsewhere in the generator and generated code already (see for example [here](https://github.com/Azure/azure-service-operator/blob/main/v2/tools/generator/internal/codegen/pipeline/status_from_swagger.go#L176)). There are ~20 hits of the `if len(errs) > 0` idiom in ASO today and 18 of this one, so we're pretty evenly split in our overall usage on it right now as well.

Reply:
Evenly split! I would have _sworn_ we were mostly (or completely) using the `if len(errs) > 0` approach. 

My problem with it is that it relies on _hidden magic_ (behaviour that's not explicit nor even implied by the method name).

That said, if you prefer this approach, leave it the way it is.